src/webmon_app/reporting/report/view_util.py

```diff
@@ -249,45 +249,54 @@ def _get_rate(id_name):
-        return None
+        else:  # the cache is not valid
+            return None
 
     runs = _get_rate("run")
     errors = _get_rate("error")
 
-    # If we didn't find good rates in the cache, recalculate them
+    # If we didn't find useful information in the cache, recalculate them
     if runs is None or errors is None:
         # check for any run in the last n-hours
         time_oldest = timezone.now() - datetime.timedelta(hours=n_hours + 1)  # is +1 needed?
 
-        # only query further if there are runs for the instrument
-        have_runs = bool(DataRun.objects.filter(instrument_id=instrument_id, created_on__gte=time_oldest).count() > 0)
-        if have_runs:
-            runs = run_rate(instrument_id, n_hours=n_hours)
-        else:
-            runs = __generate_empty_rates(n_hours)
+        # only calculate if there isn't a cache for runs
+        if runs is None:
+            # only query further if there are runs for the instrument
+            have_runs = bool(
+                DataRun.objects.filter(instrument_id=instrument_id, created_on__gte=time_oldest).count() > 0
+            )
+            if have_runs:
+                runs = run_rate(instrument_id, n_hours=n_hours)
+            else:
+                runs = __generate_empty_rates(n_hours)
+
+                # cache the run rate
+                cache.set("%s_run_rate" % instrument_id.name, runs, settings.RUN_RATE_CACHE_TIMEOUT)
+
+        # only calculate if there isn't a cache for errors
+        if errors is None:
+            have_errors = bool(
+                Error.objects.filter(
+                    run_status_id__run_id__instrument_id=instrument_id, run_status_id__created_on__gte=time_oldest
+                ).count()
+                > 0
+            )
+            if have_errors:
+                errors = error_rate(instrument_id, n_hours=n_hours)
+            else:
+                errors = __generate_empty_rates(n_hours)
 
-        # same for errors
-        have_errors = bool(
-            Error.objects.filter(
-                run_status_id__run_id__instrument_id=instrument_id, run_status_id__created_on__gte=time_oldest
-            ).count()
-            > 0
-        )
-        if have_errors:
-            errors = error_rate(instrument_id, n_hours=n_hours)
-        else:
-            errors = __generate_empty_rates(n_hours)
-
-        # cache the run rate
-        cache.set("%s_run_rate" % instrument_id.name, runs, settings.RUN_RATE_CACHE_TIMEOUT)
-        cache.set(
-            "%s_error_rate" % instrument_id.name,
-            errors,
-            settings.RUN_RATE_CACHE_TIMEOUT,
-        )
+            # cache the run rate
+            cache.set(
+                "%s_error_rate" % instrument_id.name,
+                errors,
+                settings.RUN_RATE_CACHE_TIMEOUT,
+            )
 
         # add the last run view for this instrument to the cache
         cache.set("%s_rate_last_run" % instrument_id.name, last_run)
```

Comment:
Should we be setting the `timeout` here to `settings.RUN_RATE_CACHE_TIMEOUT` as well? Or set it to `None` so it never timeouts? The default is 5 minutes so when we get `last_cached_run = cache.get("%s_rate_last_run" % instrument_id.name)` it will be `None` after 5 minutes causing it to not use the cached `run_rate`/`error_rate`.